posts/2021-03-13-microprofile-config-2.0.adoc

```diff
@@ -0,0 +1,331 @@
++
+.server.xml
+[source,xml]
+----
+<featureManager>
+    <feature>microProfile-4.0</feature>
+</featureManager>
+----
++
+If you’d rather enable the individual feature instead, you’ll need to add `<feature>mpConfig-2.0</feature>` in addition to `<feature>cdi-2.0</feature>` which enables all the cool injection functionally.
+
+And just like that you’re ready for take-off!
+
+[#Getting-Up-To-Speed]
+== Getting Up To Speed
+If you’re a little unfamiliar with MicroProfile Config (no worries if you are), link:https://download.eclipse.org/microprofile/microprofile-config-2.0/microprofile-config-spec-2.0.html[the spec] is a great place to start, or the link:https://openliberty.io/docs/21.0.0.3/external-configuration.html#[Open Liberty Docs]. 
+
+In essence, MicroProfile Config allows you to define configuration values (referred to as "config property" values) in a huge range of locations known as link:https://download.eclipse.org/microprofile/microprofile-config-2.0/microprofile-config-spec-2.0.html#configsource[ConfigSources]. These values can be easily retrieved anywhere in your code with link:https://jakarta.ee/specifications/cdi/2.0/cdi-spec-2.0.html[CDI]. For example, you could define the following in your Liberty server.xml
+
+.server.xml
+[source,xml]
+----
+<variable name="port" value="9080"/>
+----
+
+And inject it into your application code with:
+[source,java]
+----
+@Inject
+@ConfigProperty(name="port")
+int port;
+----
+
+[#New-Horizons]
+== New Horizons
+Now let’s blast off into the new features in MicroProfile Config 2.0.
+
+The examples below use the config property values defined in this example META-INF/microprofile-config.properties file (A link:https://download.eclipse.org/microprofile/microprofile-config-2.0/microprofile-config-spec-2.0.html#default_configsources[default ConfigSource]) which would be found on the classpath:
+
+[[example-ConfigSource]]
+.META-INF/microprofile-config.properties
+[source]
+----
+%testing.server.host= example.test.org
+server.host=example.org
+server.port=9080
+server.url=http://${server.host}
+----
+
+[#Config-Profile]
+=== Config Profile
+link:https://download.eclipse.org/microprofile/microprofile-config-2.0/microprofile-config-spec-2.0.html#configprofile[Config Profiles] allow configuration for different environments and development stages while only one of them is active (e.g. dev, test, prod). The Config Profile is specified using the `mp.config.profile` property, which can be set in any ConfigSource. Once it is set, the corresponding properties associated with the active profile are used. The value can also be set at application start-up, for example if you’re using Maven and the link:https://github.com/OpenLiberty/ci.maven[liberty-maven-plugin] you can start your app in link:https://openliberty.io/docs/21.0.0.2/development-mode.html[Dev mode] with:
+
+ mvn liberty:dev -Dliberty.var.mp.config.profile="testing"
```

Comment:
I think you need to explain why the property is prefixed with `liberty.var.`. Maybe include a link to some docs? https://github.com/OpenLiberty/ci.maven/blob/master/docs/common-server-parameters.md#setting-liberty-configuration-with-maven-project-properties

Would using `liberty.env` be clearer?

Reply:
We discussed this. To use `liberty.env` we'd need to define `mp_config_profile="testing"` which would probably be more confusing. 

We also found that using `liberty.bootstrap.{var}`, `liberty.env.{var}`, and `liberty.jvm.{var}` replaces their respective files rather than merging with them, which is non-desirable behaviour for this guide.